csrc/multidevice/utils.cpp

```diff
@@ -225,6 +225,30 @@ void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
 }
 } // namespace
 
+void propagateShardings(Fusion* fusion) {
+  for (auto expr : fusion->exprs()) {
+    auto inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+    auto outputs = ir_utils::filterByType<TensorView>(expr->outputs());
+    std::vector<TensorView*> input_with_mesh;
+    for (auto tv : inputs) {
+      if (tv->hasDeviceMesh()) {
+        input_with_mesh.push_back(tv);
+      }
+    }
+    NVF_ERROR(
+        !input_with_mesh.empty(),
+        "At least one input requires a DeviceMesh ",
+        expr->toString());
+    std::vector<TensorView*> outputs_without_mesh;
+    for (auto tv : outputs) {
+      if (!tv->hasDeviceMesh()) {
+        outputs_without_mesh.push_back(tv);
+      }
+    }
+    shardAllLike(input_with_mesh[0], outputs_without_mesh);
```

Comment:
Will two TensorViews in a fusion have different device mesh? I understand they can be sharded differently but I failed to see a value for them having different device meshes.

Reply:
yes! pipeline parallel!

Reply:
A really simple example might be something like this where we have one add computed on device 0 and then the second add computed on device 1
```
TensorView* a = makeTensor(2);
TensoView* b = add(a, a);
TensorView* c = set(b);
TensorView* d = add(c, c);

a->setDeviceMesh({0});
b->setDeviceMesh({0});

c->setDeviceMesh({1});
d->setDeviceMesh({1});
```